classes/local/util/manager.php

```diff
@@ -207,6 +213,100 @@ protected function prepare_files_skeletons() {
+            $componentname.'name_help',
+            $componentname.'settings',
+            $componentname.'fieldset',
+            'missingidandcmid',
+            'modulename',
+            'modulename_help',
+            'modulenameplural',
+            'nonewmodules',
+            'pluginadministration',
+            'view'
+        );
+
+        foreach ($stringids as $stringid) {
+
+            $found = false;
+            if (!empty($this->recipe['strings'])) {
+                foreach ($this->recipe['strings'] as $string) {
+                    if ($string['id'] === $stringid) {
+                        $found = true;
+                        break;
+                    }
+                }
+            }
+
+            if (!$found) {
+                $this->logger->warning("String id '$stringid' not set");
+            }
+        }
+
+        $this->prepare_file_skeleton('index.php', 'php_web_file', 'mod/index');
+        $this->prepare_file_skeleton('view.php', 'view_php_file', 'mod/view');
+        $this->prepare_file_skeleton('mod_form.php', 'php_internal_file', 'mod/mod_form');
+        $this->prepare_file_skeleton('lib.php', 'lib_php_file', 'mod/lib');
+
+        if ($this->should_have('gradebook')) {
+            $gradebookfunctions = array(
+                'scale_used',
+                'scale_used_anywhere',
+                'grade_item_update',
+                'grade_item_delete',
+                'update_grades'
+            );
+            $this->files['lib.php']->add_functions($gradebookfunctions);
+
+            $this->files['lib.php']->add_supported_feature('FEATURE_GRADE_HAS_GRADE');
+            $this->prepare_file_skeleton('grade.php', 'php_web_file', 'mod/grade');
+
+            $this->files['mod_form.php']->set_attribute('has_gradebook');
+        }
+
+        if ($this->should_have('file_area')) {
+            $fileareafunctions = array(
+                'get_file_areas',
+                'get_file_info',
+                'pluginfile'
+            );
+            $this->files['lib.php']->add_functions($fileareafunctions);
+        }
+
+        $this->files['lib.php']->add_supported_feature('FEATURE_MOD_INTRO');
+
+        // 'mod/<modname>:addinstance' and 'mod/<modname>:view' capabilities are recommended.
+        if (!$this->should_have('capabilities')) {
+            $this->logger->warning('Capabilities not defined');
+        }
+
+        if (!$this->should_have('observers')) {
+            $this->logger->warning('Observers not defined');
+        }
+
+        if (!$this->should_have('upgrade')) {
+            $this->logger->warning('Upgrade feature not defined');
+        }
+
+        // TODO: implement the backup_moodle2 feature.
+        if (!$this->should_have('backup_moodle2')) {
+            $this->logger->warning('Backup_moodle2 feature not defined');
+        }
+
+        if ($this->should_have('navigation')) {
+            $this->files['lib.php']->set_attribute('has_navigation');
+        }
```

Comment:
Firstly, why do we raise a warning if optional features (like observers or upgrade) are not enabled?

With regards to the navigation, it depends on how you see the schedule. Time is running and we will need to focus on other parts such as UI, too. Maybe you can add a file like TODO.txt with a list of suggestions for future improvements like this, and move on if it works.